scene/3d/room_manager.cpp

```diff
@@ -0,0 +1,776 @@
+
+			Vector<Vector3> dummy_pts;
+			VisualServer::get_singleton()->room_add_instance(p_room->_room_rid, light->get_instance(), light->get_transformed_aabb(), dummy_pts);
+		}
+
+		MeshInstance *mi = Object::cast_to<MeshInstance>(p_node);
+		if (mi) {
+			convert_log("\tfound mesh " + mi->get_name());
+
+			// static
+			Vector<Vector3> object_pts;
+
+			AABB aabb;
+			// get the object points and don't immediately add to the room
+			// points, as we want to use these points for sprawling algorithm in
+			// the visual server.
+			if (_bound_findpoints(mi, object_pts, aabb)) {
+				// need to keep track of room bound
+				r_room_pts.append_array(object_pts);
+
+				VisualServer::get_singleton()->room_add_instance(p_room->_room_rid, mi->get_instance(), mi->get_transformed_aabb(), object_pts);
+			} // if bound found points
+		}
+	} // if not ignore
+
+	for (int n = 0; n < p_node->get_child_count(); n++) {
+		Spatial *child = Object::cast_to<Spatial>(p_node->get_child(n));
+
+		_find_statics_recursive(p_room, child, r_room_pts);
+	}
+}
+
+bool RoomManager::_convert_manual_bound(Room *p_room, Spatial *p_node) {
+	MeshInstance *mi = Object::cast_to<MeshInstance>(p_node);
+	if (!mi)
+		return false;
+
+	Vector<Vector3> points;
+	AABB aabb;
+	if (!_bound_findpoints(mi, points, aabb))
+		return false;
+
+	mi->set_culling_portal_mode(VisualInstance::PORTAL_MODE_IGNORE);
+
+	// hide bounds after conversion
+	// set to portal mode ignore?
+	mi->hide();
+
+	return _convert_room_hull(p_room, points);
+}
+
+bool RoomManager::_convert_room_hull(Room *p_room, const Vector<Vector3> &p_room_pts) {
+	if (p_room_pts.size() > 3) {
+		Geometry::MeshData md;
+		Error err = QuickHull::build(p_room_pts, md);
+		if (err == OK) {
+			// get the planes
+			for (int n = 0; n < md.faces.size(); n++) {
+				const Plane &p = md.faces[n].plane;
+				_add_plane_if_unique(p_room->_planes, p);
+			}
+
+			convert_log("\t\t\tcontained " + itos(p_room->_planes.size()) + " planes.");
+
+			// make a copy of the mesh data for debugging
+			// note this could be avoided in release builds? NYI
+			p_room->_bound_mesh_data = md;
+
+			// aabb (should actually include portals too, NYI)
+			if (p_room_pts.size()) {
+				p_room->_aabb.position = p_room_pts[0];
+				p_room->_aabb.size = Vector3(0, 0, 0);
+			}
+
+			for (int n = 1; n < p_room_pts.size(); n++) {
+				p_room->_aabb.expand_to(p_room_pts[n]);
+			}
+
+			// send bound to visual server
+			VisualServer::get_singleton()->room_set_bound(p_room->_room_rid, p_room->_planes, p_room->_aabb);
+
+			return true;
+		}
+	}
+
+	return false;
+}
+
+void RoomManager::_add_plane_if_unique(LocalVector<Plane, int32_t> &p_planes, const Plane &p) {
+	for (int n = 0; n < p_planes.size(); n++) {
+		const Plane &o = p_planes[n];
+
+		// this is a fudge factor for how close planes can be to be considered the same ...
+		// to prevent ridiculous amounts of planes
+		const float d = _plane_simplify_dist; // 0.08f
```

Comment:
Why even define this?

```suggestion
		const real_t d = _plane_simplify_dist; // 0.08f
```